### PR TITLE
CATROID-291: Refactoring

### PR DESCRIPTION
**Pull request by gPathpp**
Changes done by @ThomasSchwengler:

* Create a `EnumUtils` helper class to avoid code duplications among enums.
* Simplify readability of `NumberFormatsTest` by fixating the expected output
* `FormulaElement`
  * Move access to static variables from within functions into parameters
  * Move common functions into a `common` package to reduce code duplication
    and enhance readability
  * Add a `function` package and move all function interpretation into classes following the open-closed principle
  * Split up many functions to improve readability
  * Simplify many functions
* `Formula`
  * Simplify many functions by reducing the number of concerns per function
* `CollisionDetection`
  * Improve memory consumption by reducing the number of memory allocations in loops
* Hide direct references to android context in wrappers
* Rename functions to be more precise
* Replace some class instances with mocks in tests where possible

Changes done by @gPathpp:

* Rebase onto develop
* Fix NumberFormats

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/common/FormulaElementOperations.kt [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+            }
+            else -> 0.0
+        }
+    }
+
+    @JvmStatic
+    fun isInteger(value: Double) = value.isFinite() && value == round(value)
+
+    @JvmStatic
+    fun tryGetLookBackgroundNumber(lookData: LookData?, lookDataList: List<LookData>) =
```
Comment (DinosaurierRex) [inline catroid/src/main/java/org/catrobat/catroid/formulaeditor/common/FormulaElementOperations.kt:104]: This searches general for the index of a look, not only Background. Please rename it

File: catroid/src/main/java/org/catrobat/catroid/formulaeditor/common/FormulaElementOperations.kt [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+        lookData?.let { lookDataList.indexOf(it) + 1.0 } ?: 1.0
+
+    @JvmStatic
+    fun getLookBackgroundName(lookData: LookData?) = lookData?.name ?: ""
```
Comment (DinosaurierRex) [inline catroid/src/main/java/org/catrobat/catroid/formulaeditor/common/FormulaElementOperations.kt:108]: same as above

File: catroid/src/main/java/org/catrobat/catroid/ui/dialogs/FormulaEditorComputeDialog.java [hunk 3]
```diff
@@ -145,22 +149,40 @@ public void onAccuracyChanged(Sensor sensor, int accuracy) {
...
+			int height = computeTextView.getLineCount() * computeTextView.getLineHeight();
+			int heightMargin = (int) (height * 0.5);
+			params.width = ViewGroup.LayoutParams.MATCH_PARENT;
+			params.height = height + heightMargin;
+			computeTextView.setLayoutParams(params);
 		});
 	}
+
+	private static class AndroidStringProvider implements Formula.StringProvider {
+		private final Context context;
```
Comment (DinosaurierRex) [inline catroid/src/main/java/org/catrobat/catroid/ui/dialogs/FormulaEditorComputeDialog.java:172]: Patrick already mentioned this issue, please refactor to dont use context here for only fetching a string

File: catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/FormulaTest.java [hunk 4]
```diff
@@ -49,19 +53,23 @@
...
+@RunWith(MockitoJUnitRunner.class)
 public class FormulaTest {
 
 	private List<InternToken> internTokenList;
 
 	@Before
 	public void setUp() {
-		internTokenList = new LinkedList<InternToken>();
+		internTokenList = new LinkedList<>();
+		if (ProjectManager.getInstance() == null) {
```
Comment (DinosaurierRex) [inline catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/FormulaTest.java:66]: Please remove unused Singleton reference to Projectmanager.

File: catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/FormulaTest.java [hunk 5]
```diff
@@ -208,7 +216,8 @@ public void stringConcatenationTest() {
 		FormulaElement joinFunctionFormulaElement = new FormulaElement(ElementType.FUNCTION,
 				Functions.JOIN.name(), null, helloStringFormulaElement, worldStringFormulaElement);
 		Formula joinFormula = new Formula(joinFunctionFormulaElement);
-		String computeDialogResult = joinFormula.getResultForComputeDialog(null);
+		StringProvider stringProvider = mock(StringProvider.class);
```
Comment (DinosaurierRex) [inline catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/FormulaTest.java:219]: Please try to put duplicate code into setup if possible

File: catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/parser/ParserTest.java [hunk 6]
```diff
@@ -152,7 +129,7 @@ public void testEmptyInput() {
 
 	@Test
 	public void testFuctionalAndSimpleBracketsCorrection() {
-		List<InternToken> internTokenList = new LinkedList<InternToken>();
+		List<InternToken> internTokenList = new LinkedList<>();
 
 		internTokenList.add(new InternToken(InternTokenType.FUNCTION_NAME, Functions.ABS.name()));
 		internTokenList.add(new InternToken(InternTokenType.FUNCTION_PARAMETERS_BRACKET_OPEN));
```
Comment (DinosaurierRex) [inline catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/parser/ParserTest.java:135]: Future Refactoring Ticket (Training). Next step could be to rewrite this class to a parameterized test.

File: catroid/src/test/java/org/catrobat/catroid/test/utiltests/NumberFormatsTest.java [hunk 7]
```diff
@@ -33,35 +34,34 @@
...
+				{"0", "0"},
+				{"8", "8"},
+				{"-120", "-120"},
+				{"0.0", "0"},
+				{"0.5", "0.5"},
+				{"0.70", "0.7"},
+				{"0.1030", "0.103"},
+				{"15.050", "15.05"},
+				{"string.1900", "string.1900"},
+				{"string0.10", "string0.10"},
```
Comment (DinosaurierRex) [inline catroid/src/test/java/org/catrobat/catroid/test/utiltests/NumberFormatsTest.java:56]: Nice, check if your Implementation works exactly how you implemented it :rofl:

File: catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/FormulaTest.java (hunk 5 above)
Comment (gPathpp) [reply catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/FormulaTest.java:219]: Unfortunately, the StringProvider is not used in all of the tests (approximately 50%)

File: catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/FormulaTest.java (hunk 4 above)
Comment (gPathpp) [reply catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/FormulaTest.java:66]: Unfortunately, this 💩 is needed if you want to execute the test locally... Ask @ThomasSchwengler, he discovered this. 
Maybe it has something to do with Dagger 🤔

File: catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/parser/ParserTest.java (hunk 6 above)
Comment (gPathpp) [reply catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/parser/ParserTest.java:135]: Is there already a ticket?

**Comment by wslany** [comment]
> Checking the error-logs coming with this PR, a new crash occurs reliably when exiting from the webview from share:
> `07-10 14:56:53.280 E/chromium(27278): [ERROR:aw_browser_terminator.cc(125)] Renderer process (28815) crash detected (code -1). `
> I tried the version without this PR, and this crash does not happen.
> 
> Note that the crash only occurs from the community; it does not occur when downloading an image from the media gallery, which also is done via a webview.
> 
> Please let me know whether this should be fixed or should be ignored for now.

Hi, I just found that the crash also appears already before. So, sorry!

**Comment by wslany** [comment]
Thanks!!!

**Pull request merged**